langchain/src/document_loaders/web/notionapi.ts

```diff
@@ -114,55 +122,107 @@ export class NotionAPILoader extends BaseDocumentLoader {
     };
   }
 
-  private async loadPage(
-    page: string | PageObjectResponse
-  ): Promise<Document[]> {
-    // Check page is a page ID or a GetPageResponse
-    const [pageData, pageId] =
-      typeof page === "string"
-        ? [this.notionClient.pages.retrieve({ page_id: page }), page]
-        : [page, page.id];
+  private async loadBlock(block: BlockObjectResponse): Promise<MdBlock> {
+    return {
+      type: block.type,
+      blockId: block.id,
+      parent: await this.n2mClient.blockToMarkdown(block),
+      children: [],
+    };
+  }
 
-    const [pageDetails, pageBlocks] = await Promise.all([
-      pageData,
-      getBlockChildren(this.notionClient, pageId, null),
-    ]);
+  private async loadBlocks(
+    blocksResponse: ListBlockChildrenResponseResults
+  ): Promise<{ mdBlocks: MdBlock[]; childDocuments: Document[] }> {
+    const blocks = blocksResponse.filter(isFullBlock);
 
-    const [childPageDocuments, childDatabaseDocuments, mdBlocks] =
+    const [childPageDocuments, childDatabaseDocuments, blocksDocsArray] =
```

Comment:
This looks like potentially a lot of requests - any concerns about rate limits?

Reply:
Yes, there is a concern about rate limits with heavily nested blocks. Though I haven't hit the rate limits yet, as this is a burst it appear to scrape by...

I have realised I need to re-architecture this to deal with rate limiting the requests.

The revised architecture would be:
- Exclusively load blocks with `loadBlock(block)` and when it finds a child database or child page add their IDs to a queue.
- Processing of the queue of pages.
- Use a scheduler (like `bottleneck`) for all notionAPI requests (which will allow for retries).

I am planning on tackling this soon. I use this document loader a lot, so I am keen on improving performance.